### ISSUE TITLE
KubernetesClientException when trying to open a namespace on a cluster that is NOT running (some older cluster kept in kubeconfig probably)

### ISSUE THREAD

**Issue by odockal** [redhat-developer/intellij-kubernetes#112]
related to #124, #103

Steps:
1. ASSERT: have cluster/context in kubeconf that doesn't exist any more
1. EXEC: pick "Set as Current Cluster" in ctx menu for this cluster/context
1. EXEC: unfold cluster/context & "Namespaces" (category)

Result:
"Namespaces" shows a child node with an error (expected), but then a notification pops up in the bottom right corner of the IDE which reports an exception:
![image.png](https://images.zenhubusercontent.com/5c924840b65bca7678bd233e/211b2c75-2776-4ec9-9fdf-122f90f27769)
```
io.fabric8.kubernetes.client.KubernetesClientException: Operation: [list]  for kind: [Namespace]  with name: [null]  in namespace: [default]  failed.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:72)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:170)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:640)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:79)
	at org.jboss.tools.intellij.kubernetes.model.resource.NonNamespacedResourcesProvider.loadAllResources(NonNamespacedResourcesProvider.kt:36)
	at org.jboss.tools.intellij.kubernetes.model.resource.NonNamespacedResourcesProvider.getAllResources(NonNamespacedResourcesProvider.kt:28)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.getAllResources(ActiveContext.kt:132)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.access$getAllResources(ActiveContext.kt:58)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext$namespacedProviders$2.invoke(ActiveContext.kt:72)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext$namespacedProviders$2.invoke(ActiveContext.kt:58)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.getNamespacedProviders(ActiveContext.kt)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.watch(ActiveContext.kt:222)
	at org.jboss.tools.intellij.kubernetes.model.ResourceModel.watch(ResourceModel.kt:113)
	at org.jboss.tools.intellij.kubernetes.tree.TreeStructure$FolderDescriptor.watchResources(TreeStructure.kt:199)
	at org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController$install$1.treeExpanded(ResourceWatchController.kt:28)
	at javax.swing.JTree.fireTreeExpanded(JTree.java:2775)
	at javax.swing.JTree.setExpandedState(JTree.java:3629)
	at javax.swing.JTree.expandPath(JTree.java:2212)
	at javax.swing.plaf.basic.BasicTreeUI.toggleExpandState(BasicTreeUI.java:2286)
	at javax.swing.plaf.basic.BasicTreeUI.selectPathForEvent(BasicTreeUI.java:2395)
	at javax.swing.plaf.basic.BasicTreeUI$Handler.handleSelection(BasicTreeUI.java:3609)
	at javax.swing.plaf.basic.BasicTreeUI$Handler.mousePressed(BasicTreeUI.java:3548)
	at com.intellij.util.ui.MouseEventAdapter.mousePressed(MouseEventAdapter.java:33)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.Component.processMouseEvent(Component.java:6545)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3325)
	at com.intellij.ui.treeStructure.Tree.processMouseEvent(Tree.java:409)
	at java.awt.Component.processEvent(Component.java:6313)
	at java.awt.Container.processEvent(Container.java:2237)
	at java.awt.Component.dispatchEventImpl(Component.java:4903)
	at java.awt.Container.dispatchEventImpl(Container.java:2295)
	at java.awt.Component.dispatchEvent(Component.java:4725)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4889)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4523)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4467)
	at java.awt.Container.dispatchEventImpl(Container.java:2281)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4725)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:737)
	at java.awt.EventQueue$4.run(EventQueue.java:735)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:719)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:664)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:363)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.net.NoRouteToHostException: No route to host (Host unreachable)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:129)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:247)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:258)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.BackwardsCompatibilityInterceptor.intercept(BackwardsCompatibilityInterceptor.java:134)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.ImpersonatorInterceptor.intercept(ImpersonatorInterceptor.java:68)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.HttpClientUtils.lambda$createHttpClient$3(HttpClientUtils.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257)
	at okhttp3.RealCall.execute(RealCall.java:93)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:490)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:451)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:433)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:166)
	... 61 more
```

**Comment by adietish** [comment]
This happens apparently when the server isn't up, right?

> Caused by: java.net.NoRouteToHostException: No route to host (Host unreachable)

ps. I'd be grateful for steps and more details so that I can reproduce the different bugs that you found

**Comment by odockal** [comment]
I thought that all necessary info to reproduce the issue was there... sorry.

**Comment by adietish** [comment]
I could fix most issues but cannot for a last one currently:

Steps:
1. ASSERT: have a cluster running where you're NOT logged in.
1. EXEC: Unfold "Custom Resources" (or any other category) 

Result:
An error is reported via the error in log icon in the bottom right border (no error dialog is shown)
![image](https://user-images.githubusercontent.com/25126/99969414-b207c800-2d9a-11eb-9f8c-b37851e8d96b.png)
```
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@66ac6faf rejected from java.util.concurrent.ScheduledThreadPoolExecutor@626e5c8a[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
	at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
	at java.util.concurrent.ScheduledThreadPoolExecutor.submit(ScheduledThreadPoolExecutor.java:632)
	at java.util.concurrent.Executors$DelegatedExecutorService.submit(Executors.java:678)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.scheduleReconnect(WatchConnectionManager.java:305)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.access$800(WatchConnectionManager.java:50)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onFailure(WatchConnectionManager.java:217)
	at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:571)
	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:198)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by adietish** [comment]
The reason for this is that the watch is being closed because the web socket for it is refused with an HTTP response code 403 (Forbidden). In reaction to this kubernetes-client shuts down the executor that would run the watch. Subsequent watch calls then get rejected (see exception above)

**Comment by adietish** [comment]
I corrected the summary and filed a separate case, where the user tries to list resources on a cluster that's running but where the user lacks authentication: #124

**Comment by adietish** [comment]
reopened since even after merging #121 an exception is still thrown: https://github.com/redhat-developer/intellij-kubernetes/pull/121#pullrequestreview-539106688. This exception is very much the same exception as reported in #124. We can thus expect the remaining exception to disappear with the new version of the kubernetes-client > 4.13.0

**Comment by adietish** [comment]
adding the "upstream" label to reflect that this should be fixed with a new kubernetes-client.

**Comment by adietish** [comment]
[kubernetes-client 5.0.0-beta1](https://search.maven.org/artifact/io.fabric8/kubernetes-client/5.0.0-beta-1/jar) was released on 20201221. We need a final or even better a 4.x release with the backported fix.

**Comment by adietish** [comment]
[kubernetes-client 5.0.0](https://search.maven.org/artifact/io.fabric8/kubernetes-client/5.0.0/jar) was released on 20201230. I'll test this version in intellij-tekton and verify that it would still work fine with this new client. If it does we'd be able to bump the client in intellij-kubernetes. If not we'd need a 4.x client with backported fixes.

**Comment by adietish** [comment]
Using [kubernetes-client 5.0.0](https://search.maven.org/artifact/io.fabric8/kubernetes-client/5.0.0/jar) as in PR #152 fixed this. I cannot reproduce it any more.
